### ISSUE TITLE
pyvmi extension compilation errors on Debian Jessie

### ISSUE THREAD

**Issue by Wenzel** [libvmi/libvmi#456]
Hi !

I tried to compile and install `pyvmi` on a Debian Jessie docker image and it failed with the following errors:
~~~
running build
running build_ext
building 'pyvmi' extension
creating build
creating build/temp.linux-x86_64-2.7
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -D_FORTIFY_SOURCE=2 -g -fstack-protector-strong -Wformat -Werror=format-security -fPIC -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/python2.7 -c pyvmi.c -o build/temp.linux-x86_64-2.7/pyvmi.o -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include 
pyvmi.c: In function 'pyvmi_init':
pyvmi.c:181:9: warning: format '%u' expects argument of type 'unsigned int', but argument 4 has type 'uint64_t' [-Wformat=]
         char *domidstring = malloc(snprintf(NULL, 0, "domid-%u", domid)+1);
         ^
pyvmi.c:181:9: warning: format '%u' expects argument of type 'unsigned int', but argument 4 has type 'uint64_t' [-Wformat=]
pyvmi.c:182:9: warning: format '%u' expects argument of type 'unsigned int', but argument 3 has type 'uint64_t' [-Wformat=]
         sprintf(domidstring, "domid-%u", domid);
         ^
pyvmi.c: In function 'pyvmi_get_vcpureg':
pyvmi.c:1311:13: error: incompatible types when assigning to type 'registers_t' from type 'int'
         reg = RAX;
             ^
pyvmi.c:1315:13: error: incompatible types when assigning to type 'registers_t' from type 'int'
         reg = RBX;
             ^
pyvmi.c:1319:13: error: incompatible types when assigning to type 'registers_t' from type 'int'
         reg = RCX;
             ^
...
...
...
pyvmi.c:1612:24: error: incompatible type for argument 3 of 'vmi_get_vcpureg'
     if (VMI_FAILURE == vmi_get_vcpureg(vmi(self), &value, reg, vcpu)) {
                        ^
In file included from pyvmi.c:31:0:
/usr/include/libvmi/libvmi.h:1901:10: note: expected 'reg_t' but argument is of type 'registers_t'
 status_t vmi_get_vcpureg(
          ^
pyvmi.c: In function 'pyvmi_v2pcache_flush':
pyvmi.c:1905:5: error: too few arguments to function 'vmi_v2pcache_flush'
     vmi_v2pcache_flush(vmi(self));
     ^
In file included from pyvmi.c:31:0:
/usr/include/libvmi/libvmi.h:1990:6: note: declared here
 void vmi_v2pcache_flush(
      ^
error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
~~~

Basically it's the same error everywhere:
~~~
pyvmi.c:1311:13: error: incompatible types when assigning to type 'registers_t' from type 'int'
         reg = RAX;
~~~

I don't have this issue on Fedora, which has `gcc 5.3.1`, compared to Jessie `gcc 4.9.2`
I created a Dockerfile for you to reproduce the issue:
~~~Dockerfile
FROM debian:jessie

RUN apt-get update && apt-get install -y git gcc python python-dev pkg-config build-essential \
        libtool automake check libglib2.0-dev libvirt-dev bison flex

RUN git clone https://github.com/libvmi/libvmi
WORKDIR libvmi
RUN ./autogen.sh && ./configure --prefix=/usr && make -j4 && make install
RUN cd tools/pyvmi && python setup.py build
~~~
Could you fix the `pyvmi` extension to make it compilable on Debian Jessie ?

Thanks !

**Comment by tklengyel** [comment]
There is already a PR to fix this issue: https://github.com/libvmi/libvmi/pull/454. It might actually get superseded by @smaresca's Python3 work.

**Comment by Wenzel** [comment]
The latest version of libvmi changed the interface of `vmi_init_complete`, and `pyvmi` cannot be compiled anymore, whatever the compiler version.

~~~C
pyvmi.c: In function ‘pyvmi_init_complete’:
pyvmi.c:209:24: error: too few arguments to function ‘vmi_init_complete’
     if (VMI_FAILURE == vmi_init_complete(&(vmi(self)), config)) {
                        ^
~~~
~~~C
pyvmi.c:165:24: warning: implicit declaration of function ‘vmi_init_custom’ [-Wimplicit-function-declaration]
     if (VMI_FAILURE == vmi_init_custom(&(vmi(object)), flags, vmiconfig)) {
~~~

~~~C
pyvmi.c:154:18: error: ‘VMI_AUTO’ undeclared (first use in this function)
         flags |= VMI_AUTO | VMI_INIT_COMPLETE;
                  ^
pyvmi.c:154:18: note: each undeclared identifier is reported only once for each function it appears in
pyvmi.c:154:29: error: ‘VMI_INIT_COMPLETE’ undeclared (first use in this function)
         flags |= VMI_AUTO | VMI_INIT_COMPLETE;
                             ^
pyvmi.c:157:29: error: ‘VMI_INIT_PARTIAL’ undeclared (first use in this function)
         flags |= VMI_AUTO | VMI_INIT_PARTIAL;
~~~

**Comment by aghamir** [comment]
Is there any update in this issue? Cannot compile pyvmi correctly:(

**Comment by Wenzel** [comment]
@aghamir i have a workaround with my own [wrapper](https://github.com/KVM-VMI/nitro/blob/master/nitro/libvmi.py) using `ctypes` in Python 3.

I just tested some of the functions that i really need, but it may help you while pyvmi is not available.

**Comment by tklengyel** [comment]
Yea, pyvmi is broken at the moment, if you need to use it, it is recommended to use the latest release of LibVMI or the wrapper by @Wenzel. We should have an official rework of it soon (pinging @smaresca).

**Comment by Wenzel** [comment]
Closing this issue as #532 has been merged.
